### DISCUSSION TITLE
Go to Symbol from Outline Panel?

### DISCUSSION THREAD

**Discussion by JamieMason** [zed-industries/zed#19782] [Feature Requests]
Hey, thanks for such a great Editor. I have a quick question please.

1. Open a file containing some code
2. Hit `cmd+shift+b` to open the [Outline Panel](https://zed.dev/docs/outline-panel)
3. Use the arrow keys to move to a symbol

How can I go to that symbol and begin editing it? The best I've managed is using `g space` which will scroll to it, but when using eg. `cmd+k cmd+→` to focus the code buffer, the cursor is still elsewhere in the file.

I've tried `Enter`, `cmd+Enter` etc but can't figure it out.

https://github.com/user-attachments/assets/3f7470de-3bcd-406f-93b8-51b5f439e06b

Thanks a lot

**Comment by SomeoneToIgnore** [comment]
There's no default binding right now, but something like 
```json
"shift-space": ["outline_panel::Open", { "change_selection": true }],
```

could be bound additionally.

I am not sure whether we should add that binding by default, or change the navigation in the outline panel entirely, navigating to each place when the outline panel entry is selected, and placing the caret there on "open".

Do you have any opinion?

**Reply by JamieMason** [reply]
Thanks a lot, adding that to my keymap has solved my problem 🎉 

> Do you have any opinion?

My intuition was that it would behave the same way as the project panel, where pressing `Enter` opens the file and focuses on its buffer. Rightly or wrongly, I expected that pressing `Enter` over eg `has_same_key_order` in my video would scroll to that function and focus on the file buffer, with the cursor on that line.

**Reply by aleclarson** [reply]
I agree with @JamieMason. Is there an issue tracking this, or is discussions the right place? I see that this discussion is closed, which might make it easy to forget about.

**Reply by SomeoneToIgnore** [reply]
https://github.com/zed-industries/zed/pull/20385 had landed into Preview and has this all.

**Reply by JamieMason** [reply]
Have there been any recent changes in this area please @SomeoneToIgnore? The keymap above is no longer working for me (Zed 0.171.6). Thanks

**Reply by SomeoneToIgnore** [reply]
There were indeed, sorry for breakage: https://github.com/zed-industries/zed/commit/176314bfd2b4ee92894be5d0555fb57443fcecd3 reworked the action and strip `change_selection` off entirely, and added a default binding to the action: https://github.com/zed-industries/zed/blob/cd34846d4b2f1d9b6f53bd5119f5ab09f82ad23c/assets/keymaps/default-macos.json#L675

The action semantics had changed somehow too, with the keyboard navigation story in the panel: now every item related to the editor open (not a directory), selected in the panel, always scrolls there, `space` places the selection to the editor and moves the focus there.

**Reply by JamieMason** [reply]
I'm seeing the editor scroll when I move up and down in the outline panel, but hitting `space` over a symbol isn't focusing it in the Editor for me. I also tried copying the default keymap entry you shared, into my keymap, but that didn't seem to work either.

EDIT: I've now switched from using `cmd+shift+b` to `cmd+shift+o` instead which is working much better and actually feels faster given the sidebar doesn't need to pop open. I think there may be a bug with the sidebar outline panel though unless I'm just misunderstanding the new behviour.

<details>
<summary>keymap.json</summary>

```json
[
	{
		"context": "Editor",
		"bindings": {
			"alt-s": "editor::SortLinesCaseInsensitive",
			"alt-u": "editor::UniqueLinesCaseSensitive",
			"alt-q": "editor::Rewrap",
			"alt-shift-down": "editor::DuplicateLineDown",
			"alt-shift-f": "editor::Format",
			"alt-shift-i": "editor::SplitSelectionIntoLines",
			"alt-shift-up": "editor::DuplicateLineUp",
			"cmd cmd": "editor::ShowCompletions",
			"cmd-\\": "pane::SplitRight"
		}
	},
	{
		"context": "Pane",
		"bindings": {
			// We use "Pane" as this is "Send" in the Assistant
			"cmd-enter": "editor::NewlineBelow"
		}
	}
]
```

</details>

<details>
<summary>settings.json</summary>

```json
{
	"assistant": {
		"default_model": {
			"provider": "copilot_chat",
			"model": "gpt-4o"
		},
		"version": "2"
	},
	"base_keymap": "VSCode",
	"buffer_font_features": {
		"calt": false
	},
	"buffer_font_size": 15,
	"code_actions_on_format": {
		"source.organizeImports.biome": true,
		"source.fixAll.biome": true
	},
	"confirm_quit": true,
	"ensure_final_newline_on_save": true,
	"experimental.theme_overrides": {
		"editor.active_line_number": "#D889FF",
		"editor.active_wrap_guide": "#FFFFFF30",
		"editor.line_number": "#FFFFFF66",
		"editor.wrap_guide": "#FFFFFF30"
	},
	"formatter": {
		"language_server": {
			"name": "biome"
		}
	},
	"hard_tabs": false,
	"preview_tabs": {
		"enabled": true,
		"enable_preview_from_file_finder": false
	},
	"project_panel": {
		"auto_fold_dirs": false
	},
	"relative_line_numbers": true,
	"remove_trailing_whitespace_on_save": true,
	"show_whitespaces": "all",
	"tab_size": 2,
	"terminal": {
		"copy_on_select": true,
		"dock": "right",
		"toolbar": {
			"title": false
		}
	},
	"theme": {
		"mode": "system",
		"dark": "Ayu Dark",
		"light": "Ayu Light"
	},
	"ui_font_family": "Departure Mono",
	"ui_font_size": 15,
	"vim_mode": true,
	"vim": {
		"use_system_clipboard": "on_yank"
	},
	"wrap_guides": [80, 100, 120]
}
```

</details>

**Reply by SomeoneToIgnore** [reply]
Seems that a bug indeed, it seems beneficial to have a single binding for opening things.

Right now, the analogue of this is `alt-enter`
https://github.com/zed-industries/zed/blob/20799c3515ae1f6c425d4b4b6c75647d85ece2b1/assets/keymaps/default-macos.json#L678-L679

but IMO this is odd, and we need to move the corresponding functionality into `Open` action.

**Reply by SomeoneToIgnore** [reply]
https://github.com/zed-industries/zed/pull/24535 should improve the situation, and `space` will open outline and search items now, togging the rest of the items (directories, files, excerpt headers, etc. — those can still be open with the `alt-enter` bindings)

**Reply by JamieMason** [reply]
Thanks a lot @SomeoneToIgnore, `alt-enter` is working as expected for me on Zed 0.172.10.
